crates/accelerate/src/quantum_circuit/circuit_data.rs

```diff
@@ -0,0 +1,365 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2023
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::quantum_circuit::circuit_data::SliceOrInt::{Int, Slice};
+use crate::quantum_circuit::intern_context::{BitType, IndexType, InternContext};
+use hashbrown::HashMap;
+use pyo3::basic::CompareOp;
+use pyo3::exceptions::{PyIndexError, PyValueError};
+use pyo3::prelude::*;
+use pyo3::types::{IntoPyDict, PySlice};
+use pyo3::{intern, PyObject, PyResult};
+use std::cmp::max;
+use std::iter::zip;
+use std::mem::swap;
+
+#[pyclass(sequence, module = "qiskit._accelerate.quantum_circuit")]
+#[derive(Clone, Debug)]
+pub struct CircuitData {
+    data: Vec<(PyObject, IndexType, IndexType)>,
+    intern_context: Py<InternContext>,
+    new_callable: PyObject,
+    fn_idx_to_qubit: PyObject,
+    fn_idx_to_clbit: PyObject,
+    fn_qubit_to_idx: PyObject,
+    fn_clbit_to_idx: PyObject,
+}
+
+#[derive(FromPyObject)]
+pub struct ElementType {
+    operation: PyObject,
+    qubits: Vec<PyObject>,
+    clbits: Vec<PyObject>,
+}
+
+#[derive(FromPyObject)]
+pub enum SliceOrInt<'a> {
+    Slice(&'a PySlice),
+    Int(isize),
+}
+
+#[pymethods]
+impl CircuitData {
+    #[new]
+    pub fn new(
+        py: Python<'_>,
+        intern_context: Py<InternContext>,
+        new_callable: PyObject,
+        qubits: PyObject,
+        clbits: PyObject,
+        qubit_indices: PyObject,
+        clbit_indices: PyObject,
+    ) -> PyResult<Self> {
+        let fn_get_item = intern!(py, "__getitem__");
+        Ok(CircuitData {
+            new_callable,
+            intern_context,
+            data: Vec::new(),
+            fn_idx_to_qubit: qubits.getattr(py, fn_get_item)?,
+            fn_idx_to_clbit: clbits.getattr(py, fn_get_item)?,
+            fn_qubit_to_idx: qubit_indices.getattr(py, fn_get_item)?,
+            fn_clbit_to_idx: clbit_indices.getattr(py, fn_get_item)?,
+        })
+    }
+
+    pub fn __len__(&self) -> usize {
+        self.data.len()
+    }
+
+    // Note: we also rely on this to make us iterable!
```

Comment:
Off the top of my head, I can't think of a clear benefit to doing so. I mostly just put this comment in because I was surprised that iteration was working out of the box (didn't know Python gives this for free), and figured it might be useful for future contributors.

Reply:
Oh, it's pretty horrendous that Python gives it for free, but in this case with Rust's ownership model, it does make things a little easier for us.  Look at this monstrosity:
```python
In [1]: class CustomGetter:
   ...:     def __init__(self, *mappings):
   ...:         self._keys, self._values = zip(*mappings)
   ...:     def __getitem__(self, index):
   ...:         # Just for illustration.
   ...:         try:
   ...:             return self._values[self._keys.index(index)]
   ...:         except ValueError:
   ...:             raise KeyError(index) from None
   ...: obj = CustomGetter(("hello", "world"), ("a", "b"), ("x", "y"))

In [2]: obj["hello"]
Out[2]: 'world'

In [3]: list(obj)
[]
```

If you implement `__getitem__` for anything _other_ than continuous sequences of ints starting from 0, the automatic iterator is nonsensical.

Reply:
Oh gosh lol. Yeah, I read somewhere that this is a left-over from a much older version of Python. But yeah, since it works without needing a custom Rust iterator struct (and all of the ownership issues that can go along with it), I figured it's good as is.